src/main/java/de/hpi/datastreams/processors/WorkerSamplingProcessor.java

```diff
@@ -2,41 +2,70 @@
 
 import de.hpi.datastreams.apps.App;
 import de.hpi.datastreams.messages.DataMessage;
+import de.hpi.datastreams.messages.MyArrayList;
+import org.apache.kafka.streams.KeyValue;
 import org.apache.kafka.streams.processor.AbstractProcessor;
 import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.state.KeyValueIterator;
 import org.apache.kafka.streams.state.KeyValueStore;
 
+import java.util.Map;
+
 public class WorkerSamplingProcessor extends AbstractProcessor<Long, DataMessage> {
 
-    private KeyValueStore<Long, DataMessage> inputDataBuffer;
+    private KeyValueStore<Long, MyArrayList<Map<Integer, Float>>> inputDataBuffer;
```

Comment:
Previously we had a `DataMessage` as value. Wouldn't this mean that we now want a `MyArrayList<DataMessage>` as value?

Reply:
As a DataMessage contains several information that we do not need (like `vectorClock` or `keyRange`) we just  have to keep the essential information (`values`) of the DataMessage. Keep in mind here, that the training data is split horizontally so the `keyRange` always contains all feature ids.
So we can simply reduce memory consumption by just storing `key -> value` pairs of the input data.